tests/test_image_untagger.py

```diff
@@ -0,0 +1,349 @@
+    assert untagger.host == host[:-1]
+    assert not untagger.remove_last
+    mock_quay_client.assert_not_called()
+    mock_quay_api_client.assert_called_once_with(token, host[:-1])
+    assert untagger._quay_client is None
+
+    client = quay_client.QuayClient("user", "pass")
+    untagger.set_quay_client(client)
+    assert client == untagger._quay_client
+
+
+@mock.patch("pubtools._quay.image_untagger.QuayClient")
+@mock.patch("pubtools._quay.image_untagger.QuayApiClient")
+def test_repo_tag_mapping(mock_quay_api_client, mock_quay_client):
+    references = [
+        "stage.quay.io/name/repo1:1",
+        "stage.quay.io/name/repo1:2",
+        "stage.quay.io/name/repo2:2",
+        "stage.quay.io/name2/repo1:1",
+        "stage.quay.io/name2/repo1:3",
+    ]
+    token = "some-token"
+    quay_user = "some-user"
+    quay_password = "some-password"
+    host = "stage.quay.io/"
+    untagger = image_untagger.ImageUntagger(
+        references, token, False, quay_user, quay_password, host
+    )
+
+    repo_tag_mapping = untagger.get_repository_tags_mapping()
+    assert repo_tag_mapping == {
+        "name/repo1": ["1", "2"],
+        "name/repo2": ["2"],
+        "name2/repo1": ["1", "3"],
+    }
+
+
+def test_tag_digest_mappings(
+    repo_api_data,
+    manifest_list_data,
+    common_tag_digest_mapping,
+    common_digest_tag_mapping,
+):
+    references = [
+        "stage.quay.io/name/repo1:1",
+        "stage.quay.io/name/repo1:2",
+        "stage.quay.io/name/repo1:3",
+        "stage.quay.io/name/repo1:4",
+    ]
+    token = "some-token"
+    quay_user = "some-user"
+    quay_password = "some-password"
+    host = "stage.quay.io/"
+    untagger = image_untagger.ImageUntagger(
+        references, token, False, quay_user, quay_password, host
+    )
+    with requests_mock.Mocker() as m:
+        m.get(
+            "https://stage.quay.io/api/v1/repository/name/repo1",
+            json=repo_api_data,
+        )
+        m.get(
+            "https://stage.quay.io/v2/name/repo1/manifests/sha256:8a3a33cad0bd33650ba7287a7ec94327d8e47ddf7845c569c80b5c4b20d49d36",
+            json=manifest_list_data,
+            headers={
+                "Content-Type": "application/vnd.docker.distribution.manifest.list.v2+json"
+            },
+        )
+        tag_digest_mapping, digest_tag_mapping = untagger.construct_tag_digest_mappings(
+            "name/repo1"
+        )
+        expected_tag_digest_mapping = common_tag_digest_mapping
+        expected_digest_tag_mapping = common_digest_tag_mapping
+
+        sort_dictionary_sortable_values(tag_digest_mapping)
+        sort_dictionary_sortable_values(expected_tag_digest_mapping)
+        sort_dictionary_sortable_values(digest_tag_mapping)
+        sort_dictionary_sortable_values(expected_digest_tag_mapping)
+
+        assert tag_digest_mapping == expected_tag_digest_mapping
+        assert digest_tag_mapping == expected_digest_tag_mapping
+        assert m.call_count == 3
```

Comment:
what is the third one?

Reply:
In repo_api_data fixture tags 1 and 2 are manifest lists (with the same digest). So the requests are: get repo data, get manifest list of tag 1, get manifest list of tag 2.

Reply:
I see